### PR TITLE
Support Extended Protection for Authentication (Channel binding)

### LINKED ISSUE #27: support Extended Protection for Authentication (Channel Binding Tokens)
Currently Organizatrions that have a `Channel Binding Token Hardening Level` domain policy  set to "Strict" ("relaxed" is the default) are unable to authenticate to the WinRM service when using NTLM over SSL (note this has always been the case and is not a recent regression). This is because the `strict` policy requires a "channel binding token" which binds the outer transport channel (TLS in this case) to the inner channel (NTLM) and povides extra protection against "man in the middle" attacks.

For more info on this topic see:
https://msdn.microsoft.com/en-us/library/dd639324.aspx
https://datatracker.ietf.org/doc/rfc5056/?include_text=1

Logistically this involves embedding a hash of the X509 cert (usually SHA-256) in the `target_info` of the `type3` message sent to authenticate the client to the server. [This post](http://blogs.msdn.com/b/openspecification/archive/2013/03/26/ntlm-and-channel-binding-hash-aka-exteneded-protection-for-authentication.aspx) does a good job clearly explaining how to create the token.

I have managed to get a valid token generated using the above technique and hacked the `session` class here to validate that it works. Here is the hack:
```
      def blob
        @blob ||=
          begin
            b = Blob.new
            b.timestamp = timestamp
            b.challenge = client_challenge
            b.target_info = inject_cbt
            b.serialize
          end
      end

      def inject_cbt
        cbt = "\x0A\x00\x10\x00\x04\x0E\x56\x28\xEC\x4A\x98\x29\x91\x70\x73\x62\x03\x7B\xB2\x3C".force_encoding(Encoding::ASCII_8BIT) #hard coded valid `MsvAvChannelBindings` using a self signed cert on a test VM

        target_info = challenge_message.target_info
        target_info.insert(target_info.length-4, cbt)

        target_info
      end
```
This resulted in succesfully authenticating over SSL to a VM with a `Channel Binding Token Hardening Level` set to `strict`.

I am starting a PR to this gem with a less "hacky" implementation and want to introduce a high level design for discussion:

* Add a optional `channel_binding_token` to `Net::NTLM::Client::Session.init_context` which would be a `ChannelBinding` instance (see next item).
* Add a `Net::Ntlm::ChannelBinding` class with a static `create` method that takes a `options` hash. There are different kinds of "outer channels" and even different kinds `tls` based tokens so one can imagine perhaps different subclasses at some point. This initial implementation would just take a `X509` cert and be capable of generating a `gss_channel_bindings_struct` and a `MsvAvChannelBindings` MD5 byte stream.
* Add a `TargetInfo` class that can be initialized from target info received in the `type2` message and have the ability to add additional `AV_PAIR` structures. The `to_s` method of this class will dump a string representation that can be fed to a `Blob`'s `target_info` accessor.

With this implementation, we can add a trivial change to the WinRM gem that would use its the SSL cert to create a `ChannelBinding` and feed that to the `Ntlm::Session` and then successfully authenticate to servers with a strict policy.

cc @sneal

### PR DESCRIPTION
**Pull request by mwrock**
fixes #27

Adds support for Extended Protection for Authentication biinding TLS channel to NTLM authentication.

### CODE REVIEW

File: lib/net/ntlm/client/session.rb [hunk 1]
```diff
@@ -213,11 +213,22 @@ def blob
...
             b.challenge = client_challenge
-            b.target_info = challenge_message.target_info
+            b.target_info = inject_cbt
             b.serialize
+            # require 'pry'
+            # binding.pry
           end
       end
 
+      def inject_cbt
```
Comment (mwrock) [inline lib/net/ntlm/client/session.rb]: a temporary hack

File: lib/net/ntlm/client/session.rb [hunk 2]
```diff
@@ -213,11 +213,22 @@ def blob
...
+        cbt = "\x0A\x00\x10\x00\x04\x0E\x56\x28\xEC\x4A\x98\x29\x91\x70\x73\x62\x03\x7B\xB2\x3C".force_encoding(Encoding::ASCII_8BIT)
```
Comment (mwrock) [inline lib/net/ntlm/client/session.rb]: Note: this is the 2 byte `AV_PAIR` id for `MsvAvChannelBindings` (0x0A) followed by the 2 byte length of the value (16 bytes or 0x10) followed by the actual binding token (same as used in the spec).

**Comment by mwrock** [comment]
This is complete. I'm gonna keep it WIP until I have a PR ready that consumes it in WinRM. I have a "hacked" WinRM using this and working against a VM with its `CbtHardeningLevel` set to `Strict`.

Tests are currently failing for 1.8.7. That was [retired](https://www.ruby-lang.org/en/news/2013/06/30/we-retire-1-8-7/) almost 3 years ago so I'll remove it in this PR.

**Comment by mwrock** [comment]
PR submitted for WinRM - https://github.com/WinRb/WinRM/pull/186 

Before removing WIP, I'm gonna do a sanity check against 2008 (R1) later. Once that checks out I think this will be good to go.

**Comment by mwrock** [comment]
Works back to Windows 2008 SP2. No longer a work in progress and ready for review.

cc @zenchild @sneal

**Comment by zenchild** [comment]
Looks good to me. Ship it :)

**Comment by mwrock** [comment]
sweet. thanks @zenchild ! squashing. Will release soon.

**Comment by sneal** [comment]
:shipit:

File: lib/net/ntlm/exceptions.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+module Net
+  module Ntlm
```
Comment (jlee-r7) [inline lib/net/ntlm/exceptions.rb]: This should be `NTLM` to match the rest of the code.

Comment (zenchild) [inline lib/net/ntlm/exceptions.rb]: Good catch @jlee-r7

Comment (mwrock) [inline lib/net/ntlm/exceptions.rb]: ah yes. thanks for catching @jlee-r7 !

**Pull request merged**